firefox-ios/Client/Frontend/Browser/Toolbars/Models/NavigationToolbarContainerModel.swift

```diff
@@ -17,8 +17,18 @@ struct NavigationToolbarContainerModel: Equatable {
     init(state: ToolbarState, windowUUID: WindowUUID) {
         self.displayBorder = state.navigationToolbar.displayBorder
         self.actions = state.navigationToolbar.actions.map { action in
-            ToolbarElement(
+            let badgeImageName: String? = if state.isPrivateMode && action.canShowPrivateBrowsingBadge {
+                action.badgeImageName
+            } else {
+                if  state.isLoginActionRequired && action.canShowWarningBadge {
+                    action.badgeImageName
+                } else {
+                    nil
+                }
+            }
+            return ToolbarElement(
                 iconName: action.iconName,
+                badgeImageName: badgeImageName,
```

Comment:
The `badgeImageName` should come from the action (and should be set through the middleware). It should not be calculated here.